File: backend-nest/src/video/video.controller.ts
import {
  Controller,
  Get,
  Param,
  Query,
  Req,
  Res,
  UseGuards,
  BadRequestException,
} from '@nestjs/common';
import { Request, Response } from 'express';
import { VideoService } from './video.service';
import { PresignService } from './presign.service';
import { JwtAuthGuard } from '../auth/jwt-auth.guard';
import {
  ApiBearerAuth,
  ApiOkResponse,
  ApiOperation,
  ApiParam,
  ApiProperty,
  ApiQuery,
  ApiTags,
} from '@nestjs/swagger';

class VideoFileDto {
  @ApiProperty({ example: 'sample.mp4' })
  id: string;

  @ApiProperty({ example: 'sample.mp4' })
  label: string;

  @ApiProperty({ example: 'video/mp4' })
  mime: string;

  @ApiProperty({ example: 'sample.mp4' })
  streamId: string;
}

class VideoListResponseDto {
  @ApiProperty({ type: [VideoFileDto] })
  files: VideoFileDto[];
}

class PresignedUrlResponseDto {
  @ApiProperty({
    example:
      'http://localhost:3000/video/stream/sample.mp4?token=abc&exp=12345678',
  })
  url: string;

  @ApiProperty({ example: 1717171717 })
  expiresAt: number;

  @ApiProperty({ example: 'sample.mp4' })
  streamId: string;
}

@ApiTags('video')
@Controller('video')
export class VideoController {
  constructor(
    private readonly videoService: VideoService,
    private readonly presignService: PresignService,
  ) {}

  @UseGuards(JwtAuthGuard)
  @Get('presign/:id')
  @ApiBearerAuth()
  @ApiOperation({
    summary: 'Generate presigned streaming URL for a given asset identifier',
  })
  @ApiParam({ name: 'id', description: 'Requested video id or filename' })
  @ApiOkResponse({ type: PresignedUrlResponseDto })
  generatePresignedUrl(@Param('id') id: string): PresignedUrlResponseDto {
    return this.presignService.generateSignedUrl(id);
  }

  @UseGuards(JwtAuthGuard)
  @Get('list')
  @ApiBearerAuth()
  @ApiOperation({ summary: 'List available video files' })
  @ApiOkResponse({ type: VideoListResponseDto })
  listVideos(): VideoListResponseDto {
    return { files: this.videoService.listAvailableVideos() };
  }

  @Get('stream/:id(*)')
  @ApiOperation({ summary: 'Stream video content using presigned URL' })
  @ApiParam({
    name: 'id',
    description: 'Resolved stream identifier (often includes path segments)',
  })
  @ApiQuery({
    name: 'token',
    description: 'HMAC token produced by presign endpoint',
    required: true,
  })
  @ApiQuery({
    name: 'exp',
    description: 'Expiration timestamp from presign endpoint',
    required: true,
  })
  @ApiOkResponse({
    description: 'Byte-range video stream',
    schema: { type: 'string', format: 'binary' },
  })
  streamVideo(
    @Param('id') id: string,
    @Query('token') token: string,
    @Query('exp') exp: string,
    @Req() req: Request,
    @Res() res: Response,
  ) {
    if (!token || !exp) {
      throw new BadRequestException('Missing token or exp');
    }

    const numericExp = Number(exp);
    if (Number.isNaN(numericExp)) {
      throw new BadRequestException('exp must be a UNIX timestamp');
    }

    this.presignService.validateSignature(id, numericExp, token);
    this.videoService.streamVideo(id, req, res);
  }
}

File: backend-java/src/main/java/com/streaming/features/video/VideoController.java
package com.streaming.features.video;

import com.streaming.features.video.models.PresignedUrlResponseDto;
import com.streaming.features.video.models.VideoListResponseDto;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.io.IOException;

@RestController
@RequestMapping("/video")
@Tag(name = "video", description = "Video catalog and streaming endpoints")
public class VideoController {

    private final VideoService videoService;
    private final PresignService presignService;

    public VideoController(VideoService videoService, PresignService presignService) {
        this.videoService = videoService;
        this.presignService = presignService;
    }

    @GetMapping("/presign/{id}")
    @SecurityRequirement(name = "bearerAuth")
    @Operation(summary = "Generate presigned streaming URL for a given asset identifier")
    public PresignedUrlResponseDto generatePresignedUrl(
            @Parameter(description = "Requested video id or filename") @PathVariable String id) {
        var result = presignService.generateSignedUrl(id);
        return new PresignedUrlResponseDto(
                (String) result.get("url"),
                (long) result.get("expiresAt"),
                (String) result.get("streamId")
        );
    }

    @GetMapping("/list")
    @SecurityRequirement(name = "bearerAuth")
    @Operation(summary = "List available video files")
    public VideoListResponseDto listVideos() {
        return new VideoListResponseDto(videoService.listAvailableVideos());
    }

    @GetMapping("/stream/{*id}")
    @Operation(summary = "Stream video content using presigned URL",
            responses = {
                    @ApiResponse(responseCode = "206", description = "Byte-range video stream",
                            content = @Content(mediaType = "application/octet-stream", schema = @Schema(type = "string", format = "binary"))),
                    @ApiResponse(responseCode = "200", description = "Full video stream")
            })
    public ResponseEntity<Resource> streamVideo(
            @Parameter(description = "Resolved stream identifier (often includes path segments)") @PathVariable("id") String id,
            @Parameter(description = "HMAC token produced by presign endpoint", required = true) @RequestParam(value = "token", required = false) String token,
            @Parameter(description = "Expiration timestamp from presign endpoint", required = true) @RequestParam(value = "exp", required = false) String exp) {

        // Remove leading slash that Spring Boot adds for wildcard paths depending on version
        if (id != null && id.startsWith("/")) {
            id = id.substring(1);
        }

        if (token == null || exp == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Missing token or exp");
        }

        long numericExp;
        try {
            numericExp = Long.parseLong(exp);
        } catch (NumberFormatException e) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "exp must be a UNIX timestamp");
        }

        presignService.validateSignature(id, numericExp, token);

        return videoService.streamVideo(id);
    }
}
